consensus/istanbul/backend/backend.go

```diff
@@ -497,9 +492,16 @@ func (sb *Backend) Commit(proposal istanbul.Proposal, aggregatedSeal types.Istan
 		return nil
 	}
 
-	if sb.broadcaster != nil {
-		sb.broadcaster.Enqueue(fetcherID, block)
+	// Note that result should NOT be nil, one edge case is when the validator re-start in the middle of a sequence.
+	// When happening, we re-verify the proposal and write it.
+	if result == nil {
+		sb.logger.Error("Verifying the block again", "number", block.Number(), "hash", block.Hash())
+		if result, _, err = sb.Verify(proposal); err != nil {
+			sb.logger.Info("Verified proposal", "err", err)
+			return err
+		}
 	}
+	go sb.onNewConsensusBlock(block, result.Receipts, result.Logs, result.State)
```

Comment:
So now `onNewConsensusBlock` is calling `blockchain.WriteBlockWithState` which as far as I know does not broadcast the block (doesn't result in a call to `ProtocolManager.BroadcastBlock`), so I'm wondering how does the block get propagated to peers in this case, and how does that affect performance?

Reply:
It does now:

```go
func(block *types.Block, receipts []*types.Receipt, logs []*types.Log, state *state.StateDB) {
  if err := w.chain.InsertPreprocesedBlock(block, receipts, logs, state); err != nil {
	  if err == core.ErrNotHeadBlock {
		  log.Warn("Tried to insert duplicated produced block", "blockNumber", block.Number(), "hash", block.Hash(), "err", err)
	  } else {
		  log.Error("Failed to insert produced block", "blockNumber", block.Number(), "hash", block.Hash(), "err", err)
	  }
	  return
  }
  log.Info("Successfully produced new block", "number", block.Number(), "hash", block.Hash())
  
  if err := w.mux.Post(core.NewMinedBlockEvent{Block: block}); err != nil {
	  log.Error("Error when posting NewMinedBlockEvent", "err", err)
  }
  })
```